libraries/WiFi/src/WiFiGeneric.h

```diff
@@ -141,6 +142,7 @@ static const int WIFI_SCANNING_BIT = BIT11;
 static const int WIFI_SCAN_DONE_BIT= BIT12;
 static const int WIFI_DNS_IDLE_BIT = BIT13;
 static const int WIFI_DNS_DONE_BIT = BIT14;
+static const int WIFI_WANT_IP6_BIT = BIT15;
```

Comment:
Hmm, I can see this is an ETH_HAS_IP_BIT, and ETH_HAS_IP6_BIT ... so maybe we need some flags about whether we have been allocated an address of the relevant type (not just if we want one).

I don't know if HAS refers to an assigned address (e.g. prefix from a RA, DHCP assignment, etc), or if a self-assigned link-local address is sufficient.

Probably need a way to get localIPs() for all your local IPs and determine if you have a global IPv6, or only a link-local, or none at all.

Reply:
exactly. this only means that we enabled IPv6 support, not that we have an address present.

Reply:
When you connect though, if WANT is on, then you do generate a link-local address, so you will usually have one of them ... but it is not very useful for connecting to a public address or anything. Maybe once this is merged, then next thing is to support returning multiple results for things like getHostsByName() and getLocalIPs() to get all of them, which would allow manual usage of IPv6, while the default (single return) is worked out.

Reply:
and then there is Ethernet, which does not raise `WIFI_WANT_IP6_BIT`. Maybe we need bits that tell the actual state, like `NET_HAVE_IP6_LINK_LOCAL_BIT` and `NET_HAVE_IP6_GLOBAL_BIT` to signal when we can serve V6 queries. DNS would return GLOBAL and mDNS would return LINK_LOCAL if I am not mistaken.

Reply:
But you do get an event telling you received an IPv6 address.
Actually, you may even get 2 (a few seconds inbetween) if you also receive a global scope one.

Not sure if you can always expect DNS to return global, as some routers allow to serve local DNS queries based on the hostname.
For example Fritzbox can perfectly fine resolve just based on the hostname and Mikrotik often (by default) can resolve <hostname>.lan queries.
Though the Fritzbox only seems to do this on those who are set to get a static lease (as in always the same IP on a DHCP request)
A quick test here always returned the global scope address, as all those with static leases also have a global scope IPv6 address.
But at least it does show that the global scope address is the preferred answer from the DNS server (which makes sense)

Reply:
based on the received address, it can be determined wether it's global or local. It is easy to check the address from DNS as well. Overall I think we can leave such things for later. I have mikrotiks here. For V6 DNS to function properly for the general case (you want to connect to remote server with global v6 address) we need to make sure that we have global address, before we attempt to ask for v6 addresses from DNS. What I also found, is that we need to track when v6 and v4 become available or not and clear the dns cache, else we will always get a v4 address if it's already in the cache.

BTW I am able currently to connect to google using my global address and v6 dns
```
found 2a00:1450:4017:814::200e

connecting to google.com
HTTP/1.1 301 Moved Permanently
Location: http://www.google.com/
Content-Type: text/html; charset=UTF-8
Content-Security-Policy-Report-Only: object-src 'none';base-uri 'self';script-src 'nonce-vvtVgn_tG0NhL5a6_AT5KQ' 'strict-dynamic' 'report-sample' 'unsafe-eval' 'unsafe-inline' https: http:;report-uri https://csp.withgoogle.com/csp/gws/other-hp
Date: Wed, 10 Jan 2024 11:24:51 GMT
Expires: Fri, 09 Feb 2024 11:24:51 GMT
Cache-Control: public, max-age=2592000
Server: gws
Content-Length: 219
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN

<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="http://www.google.com/">here</A>.
</BODY></HTML>
closing connection
```

Reply:
Just curious about handling IPv6 DNS servers.
So DNS servers with an IPv6 address.
I don't think it is really an immediate must have, but in the end it probably will be needed. For example when using IPv6-only networking.

Reply:
you usually get those automatically

Reply:
What I meant is I didn't see any functions to set IPv6 addresses for DNS server.

Reply:
Here is a flow with logic based on `NET_HAS_IPV6_GLOBAL_IP` that is currently running here:
```
ETH Connected

ETH Got IPv6[2]: 'eth0' (en1)
eth0: <UP,100M,FULL_DUPLEX,AUTO>
      ether 32:C6:F7:F4:2F:8C phy 0x1
      inet6 fe80::30c6:f7ff:fef4:2f8c%en1 type LINK_LOCAL

ETH Got IP[2]: 'eth0' (en1)
eth0: <UP,100M,FULL_DUPLEX,AUTO>
      ether 32:C6:F7:F4:2F:8C phy 0x1
      inet 192.168.254.22 netmask 255.255.255.0 broadcast 192.168.254.255
      gateway 192.168.254.1 dns 192.168.254.3
      inet6 fe80::30c6:f7ff:fef4:2f8c%en1 type LINK_LOCAL

connecting to google.com
dns_found 142.251.140.46
HTTP/1.1 301 Moved Permanently
Location: http://www.google.com/
......
closing connection

ETH Got IPv6[2]: 'eth0' (en1)
eth0: <UP,100M,FULL_DUPLEX,AUTO>
      ether 32:C6:F7:F4:2F:8C phy 0x1
      inet 192.168.254.22 netmask 255.255.255.0 broadcast 192.168.254.255
      gateway 192.168.254.1 dns 192.168.254.3
      inet6 fe80::30c6:f7ff:fef4:2f8c%en1 type LINK_LOCAL
      inet6 2a10:bac0:b000:7395:30c6:f7ff:fef4:2f8c type GLOBAL

connecting to google.com
[ 21443][D][WiFiGeneric.cpp:1621] hostByName(): Clearing DNS cache
dns_found 2a00:1450:4017:814::200e type 1
HTTP/1.1 301 Moved Permanently
Location: http://www.google.com/
.......
closing connection

```

Does this make sense to you? Because it does for me. If have global address, then prefer V6, else force V4

Reply:
That totally makes sense, but it is not what I meant :)

What I mean is that the IP of a DNS server (thus not a DNS request, but the server you send the DNS query to) can be IPv6.
But so far I only found code for DNS server to have IPv4 addresses.

I know an DNS server reachable at an IPv4 address is perfectly capable of answering an `AAAA` query, but how can we reach a DNS server at an IPv6 address?

Reply:
i'm pretty sure that happens behind the scenes. I do not see a way to add v6 DNS or gateway, yet it all works

Reply:
Like I said, it is likely not an issue to fix for now, but if you come accross some related code for these, keep in mind it might perhaps be useful to add. Though not sure if you really need to have it when only working on IPv6 link local.

Reply:
> DNS would return GLOBAL and mDNS would return LINK_LOCAL

No. Both return global addresses. mDNS operates in link-local scope, but returns global addresses.

A DNS server will return whatever addresses it is configured with. Also 'local' and link-local are different things.

e.g. On my local LAN I can do (DNS):
```
> host nazgul-qnap.lan
nazgul-qnap.lan has address 192.168.1.39
nazgul-qnap.lan has IPv6 address fd7c:e25e:67e8:40::1027
nazgul-qnap.lan has IPv6 address 2407:8800:bc61:1340::1027
```

Which is a local device, not available to the wider internet. I'm not sure if DNS supports link-local addresses.

mDNS is only the local LAN, i.e. multicast SCOPE is link-local), but it still returns global addresses (not link-local ADDRESSES).

```
> avahi-resolve --name nazgul.local
NAZGUL.local	2407:8800:bc61:1340::1027
> avahi-resolve --name nazgul.local -4
NAZGUL.local	192.168.1.39
```

Reply:
> didn't see any functions to set IPv6 addresses for DNS server.

This generally happens behind the scenes... you also won't see any functions to set IPv4 DNS servers.

DNS servers are configured via various means. For IPv4 they are part of DHCP parameters.

For IPv6 they can either be part of RA (RA DNS), or RA can specify to look at DHCPv6. (DHCPv6 can either the stateful -giving out addresses, or stateless - where it only gives out DNS, etc). Some devices (e.g. Android phones) only use RA DNS and don't support DHCPv6, even for DNS options.

From a quick look at LwIP, it supports stateless DHCPv6, i.e. DHCPv6 for DNS (but not addresses).

The actual DNS servers provided can be a list of both IPv6 and IPv4.

e.g. my local LAN (both are just my router)

```
   DNS Servers . . . . . . . . . . . : fd7c:e25e:67e8:40::1
                                       192.168.1.1
```

And if you connect on either of those addresses you will get the same DNS response (with both IPv4 and IPv6 addresses, as relevant).

Reply:
Yeah, so this page on LwIP: https://www.nongnu.org/lwip/2_1_x/group__lwip__opts__dhcpv6.html#ga92c3bc242ad20a5f398c45d332864a29

Says that stateful DHCPv6 (for getting addresses) is not supported yet 

But it does support stateless for DNS and any DNS servers received in DHCPv6 responses are passed to dns_setserver().  In the Neighbour Discovery section, just above, it also mentions RDNSS and getting any DNS servers advertised in IPv6 RA messages and also adding them to DNS as well.